x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/PredictLinear.java

```diff
@@ -0,0 +1,166 @@
+        this.timestamp = ts;
+        this.t = t;
+    }
+
+    private PredictLinear(org.elasticsearch.common.io.stream.StreamInput in) throws java.io.IOException {
+        super(
+            Source.readFrom((PlanStreamInput) in),
+            in.readNamedWriteable(Expression.class),
+            in.readNamedWriteable(Expression.class),
+            in.readNamedWriteable(Expression.class),
+            in.readNamedWriteableCollectionAsList(Expression.class)
+        );
+        this.t = children().get(4);
+        this.timestamp = children().get(3);
+    }
+
+    @Override
+    public Expression timestamp() {
+        return timestamp;
+    }
+
+    @Override
+    public AggregateFunction perTimeSeriesAggregation() {
+        return this;
+    }
+
+    @Override
+    public AggregateFunction withFilter(Expression filter) {
+        return new PredictLinear(source(), field(), filter, window(), timestamp, t);
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.DOUBLE;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new PredictLinear(
+            source(),
+            newChildren.get(0),
+            newChildren.get(1),
+            newChildren.get(2),
+            newChildren.get(3),
+            newChildren.get(4)
+        );
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, PredictLinear::new, field(), filter(), window(), timestamp, t);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public AggregatorFunctionSupplier supplier() {
+        if (t.foldable() == false) {
+            throw new IllegalArgumentException("The 't' parameter of the 'predict_linear' function must be a constant value.");
+        }
+        final Double timeDiffSeconds = switch (t.fold(FoldContext.small())) {
+            case Duration d -> d.toMillis() / 1000.0;
+            case Long l -> l.doubleValue();
+            case Integer i -> i.doubleValue();
+            case String s -> Duration.from(Objects.requireNonNull(EsqlDataTypeConverter.parseTemporalAmount(s, DataType.TIME_DURATION)))
+                .toMillis() / 1000.0;
+            default -> throw new IllegalArgumentException(
+                "The 't' parameter of the 'predict_linear' function must be of type long, integer, keyword, or timedelta. It was of type: "
+                    + t.dataType()
+            );
+        };
+
+        SimpleLinearRegressionWithTimeseries.SimpleLinearModelFunction fn = (SimpleLinearRegressionWithTimeseries model) -> {
+            double slope = model.slope();
+            double intercept = model.intercept();
+            double lastTs = model.lastTimestamp();
+            if (Double.isNaN(slope)) {
+                return Double.NaN;
+            }
+            return intercept + slope * (lastTs + timeDiffSeconds);
```

Comment:
I think there is a mismatch: lastTs is in milliseconds, while timeDiffSeconds is in seconds.

Reply:
omg great catch thanks